sdk/core/perf/src/program.cpp

```diff
@@ -224,50 +230,78 @@ inline void RunTests(
   for (size_t index = 0; index != tests.size(); index++)
   {
     tasks[index] = std::thread(
-        [index, &tests, &completedOperations, &lastCompletionTimes, &deadLineSeconds, &context]() {
+        [index,&progressToken, &tests, &completedOperations, &lastCompletionTimes, &deadLineSeconds, &context]() {
           bool isCancelled = false;
           // Azure::Context is not good performer for checking cancellation inside the test loop
-          auto manualCancellation = std::thread([&deadLineSeconds, &isCancelled] {
-            using namespace std::chrono_literals;
-            std::this_thread::sleep_for(deadLineSeconds);
+          auto manualCancellation = std::thread([&deadLineSeconds, &progressToken, &isCancelled] {
+            std::chrono::system_clock::time_point timeoutTime
+                = std::chrono::system_clock::now() + deadLineSeconds;
+            do
+            {
+              if (progressToken.IsCancelled())
+              {
+                return;
+              }
+              std::this_thread::sleep_for(std::chrono::milliseconds(500));
+            } while (std::chrono::system_clock::now() < timeoutTime);
             isCancelled = true;
           });
-
-          RunLoop(
-              context,
-              *tests[index],
-              completedOperations[index],
-              lastCompletionTimes[index],
-              false,
-              isCancelled);
-
-          manualCancellation.join();
+          try
+          {
+
+            RunLoop(
+                context,
+                *tests[index],
+                completedOperations[index],
+                lastCompletionTimes[index],
+                false,
+                isCancelled);
+
+            manualCancellation.join();
+          }
+          catch (std::exception const& ex)
+          {
+            std::cout << "Test failed with exception: " << ex.what() << std::endl;
+            progressToken.Cancel();
+            manualCancellation.join();
+            std::cout << "Cancelling tests." << std::endl;
+            return;
+          }
         });
   }
   // Wait for all tests to complete setUp
   for (auto& t : tasks)
   {
     t.join();
   }
+  bool testPrematurelyCancelled = progressToken.IsCancelled();
 
   // Stop progress
-  progresToken.Cancel();
+  progressToken.Cancel();
   progressThread.join();
 
-  std::cout << std::endl << "=== Results ===";
+  if (testPrematurelyCancelled)
+  {
+    std::cout << std::endl << "*** TESTS CANCELLED DUE TO EXCEPTION ***" << std::endl; 
```

Comment:
Well, I get what you are doing here...  but, the expectation is to use perf tests for catching perf-related things.
Because of that, the perf-tests should be expected to be run after unit tests and functional tests (which are the gates to find application issues (like exceptions). ).

What I want to say is that, we should not worry or take care of perf-tests which throws exceptions. Wouldn't that mean that unit tests would be failing as well?

Or am I missing something??

Reply:
The way I hit this was I accidentally mistyped the --url parameter to the WinHttpClient test. And I had no way of figuring out what went wrong when that happened.

Pretty much all of the changes I made in this PR are about handling user errors when running the perf test application, which happens a lot when the user is setting up the test.

Reply:
I wonder if there is a way to ask Windows to behave the same as Linux..  You get the error description from the exception on Linux.

Reply:
On linux, running with no arguments gives you an error in the console that list the available test names.
Then, missing required arguments or using bad input for a test name would also give a nice error description in the console.

Reply:
There is, kinda. You can suppress the message box (my most recent change does that). However without the exception handler, the code will never print out the value of the unhandled exception. And that defeats the point of throwing an exception.

Reply:
I'm asking vcsig to see if anyone there has any suggestions, but until then we're stuck with sprinkling try/catch handlers to catch this kind of stuff :(.

Reply:
And unfortunately, it appears the answer is "you can't". The best suggestion they have requires registering a per-thread terminate handler, which defeats the point (I want a one-stop solution).